compositor_pipeline/src/queue/utils.rs

```diff
@@ -0,0 +1,90 @@
+use std::{
+    collections::VecDeque,
+    mem,
+    time::{Duration, Instant},
+};
+
+use compositor_render::{AudioSamplesBatch, Frame};
+
+use super::DEFAULT_BUFFER_DURATION;
+
+#[derive(Debug)]
+pub(super) enum InputState<Payload: ApplyOffsetExt> {
+    WaitingForStart,
+    Buffering { buffer: Vec<(Payload, Duration)> },
+    Ready { offset: chrono::Duration },
+}
```

Comment:
Why should we do it in such a weird way? Can't we just keep `offset: Option<chrono::Duration>` and apply it to every enqueued frame/sample? I don't understand how this helps.

Reply:
No. I calculate offset when switching from buffering to ready state.